### ISSUE TITLE
Nested selectors behaviour

### ISSUE THREAD

**Issue by mattfysh** [emotion-js/emotion#235]
- `emotion` version: latest v7
- `react` version: latest

Relevant code.

```javascript
const section = css`
  background-color: blue;
  button {
    border-radous: 4px;
  }
`
```

What you did:

Used a nested selector, with no `&` placed within the nested selector

What happened:

The output CSS selector became `css-section-hashbutton`

Reproduction:

Problem description:

The output selector is invalid, and does not represent the nested relationship of `button` to `section`
Suggested solution:

The output selector should instead produce `css-section-hash button` (CSS child/descendent selector)

**Comment by thangngoc89** [comment]
@mattfysh This is an intended behavior in v7. `&` is a requirement now

**Comment by mattfysh** [comment]
This smells funky, and it creates a semantic rift between what it means to nest selectors and what it means to nest elements.

**Comment by thangngoc89** [comment]
@mattfysh I forgot why `&` is a requirement but this is actually how CSS nested proposal works. You can try it yourself in cssnext Playground http://cssnext.io/playground/


```
section {
 & button {
   font-size: 16px;
 }
}
```

compiles to 

```
section button {
    font-size: 16px;
}
```

**Comment by tkh44** [comment]
The `&` is replaced with the generated class name so that you get the correct specificity. This is a side effect of using objects internally, but I have some ideas on how we can get around this.

**Comment by darrentorpey** [comment]
This is very good to know. I had my nested selectors break when I updated to v7 and didn't understand why until I read this. I don't remember it being in the blog post about v7, even though it's a breaking change. Maybe it was just too subtle or I wasn't reading carefully. :)

**Comment by tkh44** [comment]
I definitely forgot to add this to the post. Sorry about that.

**Comment by darrentorpey** [comment]
No worries; I know I'm basically working with alpha tech here. :)

I think the changes in v7 are good -- esp. dealing with the horrible React warning about props and slimming things down by making `theming` optional and more of a standard / less magical use of it. It's just that I'm just now learning how this all works so some of the changes are hard to follow.

I'm working on an example app that I guess I'll throw into a temp repo to show some of the problems I've run into.

**Comment by silvenon** [comment]
Should this be enforced with component selectors as well? Or at least enabled.

```js
const A = styled.a`
  ...
`;

// this works
const B = styled.b`
  ${A} {
    ...
  }
`;
// but this doesn't
const B = styled.b`
  & ${A} {
    ...
  }
`;
```

**Comment by mattfysh** [comment]
@silvenon - be careful interpolating generated classes like that, as once you add a `composes` to `A` it will become a list of classes, which can't be rendered to a selector like that,

e.g. if adding composes to A, it becomes `A = 'css-hash-A1 css-hash-A2'`
and your selector become `css-hash-B css-hash-A1 css-hash-A2` which does not match your intended selector use case of `B A` ... it would only work with `B A A` 🐑

**Comment by hamlim** [comment]
Stumbled upon another unique edge case bug related to nested selectors, it seems you can't target a nested selector without applying styles to the containing styled element or classname:

```Javascript
// Doesn't work
const Wrapper = styled.div`
  & .${someStyles} {
    color: red;
  }
`
// Does work
const Wrapper = styled.div`
  display: block;
  & .${someStyles} {
    color: red;
  }
`
```

Don't think it would ever make sense to just target another class without some other styles applied to the parent however just something to note.

**Comment by mitchellhamilton** [comment]
@hamlim I can't seem to reproduce this. Are you using an older version of emotion?

**Comment by silvenon** [comment]
I was able to reproduce the above! This:

```js
const foo = css`
  color: blue;
`;
const Bar = styled.div`
  & .${foo} {
    color: red;
  }
`;
```

is being parsed as if I did this:

```js
const foo = css`
  color: blue;
`;
const Bar = styled.div`
  ${foo};
`;
```

but it works as expected if there is a rule before, like @hamlim said. I'm using the latest Emotion.

I just filed the bug here: #255

**Comment by goodmind** [comment]
What about reverse selector?
```js
const A = styled.div`
  color: #000;
`;

const B = styled.div`
  color: red;
  ${A} & {
    color: green;
  }
`;
```

**Comment by mitchellhamilton** [comment]
All of these issues are fixed in emotion 8, closing.
